eng/scripts/Verify-And-Merge-PRs.ps1

```diff
@@ -1,111 +1,250 @@
+    $repoOwner, $repoName, $Number = $line.Split(";")
+    $prList += @{ RepoOwner = $repoOwner; RepoName = $repoName; Number = $Number }
+  }
 
-  $RetryCount--
-  if (($unMergablePRs.Count -gt 0) -and ($RetryCount -gt 0) -and $ShouldMerge)
-  {
-    $PRData = $unMergablePRs
-    Start-Sleep -Seconds 30
-    LogDebug "Retrying merging for unmergable Prs"
-    $ReadyForMerge = $true
+  $mergeable = ProcessPRMergeStatuses $prList
+
+  if ($SkipMerge) {
+    LogInfo "Skipping merge of $($mergeable.Length) PRs due to -SkipMerge parameter."
+    return
   }
+
+  MergePRs $mergeable
 }
-while (($unMergablePRs.Count -gt 0) -and ($RetryCount -gt 0) -and $ShouldMerge)
 
+function ProcessPRMergeStatuses([array]$prData) {
+  for ($retry = 1; $retry -le 5; $retry++)
+  {
+    $curr, $prData = $prData, @()
+    foreach ($pr in ($curr | Where-Object { $_.Retry -ne $false }))
+    {
+      $prData += GetOrSetMergeablePR -repoOwner $pr.RepoOwner -repoName $pr.RepoName -prNumber $pr.Number
+    }
+
+    if ($SkipMerge -or $prData.Retry -notcontains $true) {
+      break
+    }
+
+    $sleep = [Math]::Pow(2, $retry)
+    LogInfo "Some PRs were not in a mergeable state, retrying after $sleep seconds..."
+    Start-Sleep -Seconds $sleep
+  }
 
-if (!$ReadyForMerge) {
-  LogError "At least one sync PR is not able to be merged please investigate and then retry running this job to auto-merge them again"
-  exit 1
+  if ($prData.Block -contains $true) {
+    LogError "The following sync PRs are not able to be merged. Investigate and then retry running this job to auto-merge them again"
+    $prData | Where-Object { $_.Block } | ForEach-Object { LogInfo $_.Url }
+    exit 1
+  }
+
+  return $prData
 }
 
-if ($ReadyForMerge -and $ShouldMerge)
-{
-  # Merge Pull Requests
-  foreach ($mergablePRObj in $mergablePRs)
+function MergePRs([array]$toMerge) {
+  foreach ($pr in $toMerge)
   {
-    $mergablePR = $mergablePRObj.Url
-    $mergeApiUrl = $mergablePR + "/merge"
-
-    Write-Host $mergablePRObj.HeadSHA
-    $data = @{
-      sha = $mergablePRObj.HeadSHA
+    LogInfo $toMerge.HeadSHA
+    $body = @{
+      sha = $toMerge.HeadSHA
       merge_method = "squash"
-    }
+    } | ConvertTo-Json -Compress
 
-    # Merge Pull Request
     try {
-      Write-Host "Merging $mergablePR"
-      $response = Invoke-RestMethod -Method Put -Headers $headers $mergeApiUrl -Body ($data | ConvertTo-Json)
+      LogInfo "Merging $($toMerge.MergeUrl)"
+      $response = Invoke-RestMethod -Method Put -Headers $Headers $toMerge.MergeUrl -Body $body
     }
     catch {
-      LogError "Invoke-RestMethod [$mergeApiUrl] failed with exception:`n$_"
-      exit 1
+      LogError "Invoke-RestMethod [$($toMerge.MergeUrl)] failed with exception."
+      throw
+    }
+  }
+}
+
+function RequestGithubGraphQL([string]$query, [object]$variables = @{}) {
+  $body = @{ query = $query; variables = $variables } | ConvertTo-Json -Depth 100
+
+  try {
+    $response = Invoke-RestMethod `
```

Comment:
I would suggest switching to using the gh cli if we are touching this. You can use it to call graphql.

Reply:
Example of doing it https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/scripts/Delete-RemoteBranches.ps1#L20

Reply:
Totally in favor, just wasn't sure if we wanted to stick to the default invoke-restmethod here or not. I'll make the change.

Reply:
I've been moving to the gh cli over calling rest directly in PS.